### PR TITLE
[#833][#836] Fix invitation flow

### PR DESCRIPTION
**Pull request by danielkryska**
- [Fix] After declining the invitation, the link leads to page 404 #833
- [Fix] Fix adding a user to an institution on invitation acceptance #836

### CODE REVIEW

File: s4e-backend/src/test/java/pl/cyfronet/s4e/controller/AppUserControllerTest.java [hunk 1]
```diff
@@ -175,10 +173,17 @@ public void handle(OnResendRegistrationTokenEvent event) {
...
         }
+
+        @EventListener
+        public void handle(OnConfirmInvitationEvent event) {
+
+        }
     }
 
     @Test
     public void shouldCreateUser() throws Exception {
```
Comment (kmarszalek) [inline s4e-backend/src/test/java/pl/cyfronet/s4e/controller/AppUserControllerTest.java:184]: i would make test for regiter with and wotuhout token

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java [hunk 2]
```diff
@@ -65,13 +65,21 @@
...
         validateRecaptcha(request);
 
         try {
             appUserService.register(registerRequest);
+
+            val token = registerRequest.getToken();
+            if (token != null) {
+                invitationService.confirmInvitation(email, token);
+                val confirmInvitationEvent = new OnConfirmInvitationEvent(token, LocaleContextHolder.getLocale());
+                eventPublisher.publishEvent(confirmInvitationEvent);
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java:80]: I'm not a fan of breaking the SRP in this method, however I understand the workflow. Nevertheless, add a test which verifies, that if line 78 throws NFE, then creating a new user is rolled back

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/request/RegisterRequest.java [hunk 3]
```diff
@@ -36,4 +36,6 @@
     @CountryCode
     @Schema(description = "User's ISO 3166 alpha-2 country code", example = "PL")
     private String country;
+
+    private String token;
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/request/RegisterRequest.java:40]: It's an optional part, which in other endpoints is passed as query param. I would put it in query param in this case too.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/InvitationService.java [hunk 4]
```diff
@@ -56,12 +56,12 @@ public String createInvitationFrom(String email, String institutionSlug, boolean
     }
 
     @Transactional(rollbackFor = NotFoundException.class)
-    public void confirmInvitation(String token) throws NotFoundException {
+    public void confirmInvitation(String userEmail, String token) throws NotFoundException {
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/service/InvitationService.java:59]: Great change, it's better to fetch the AppUserDetails in controller layer and inject it into the service layer.

**Review by jswk — CHANGES_REQUESTED** [review]

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java [hunk 5]
```diff
@@ -64,14 +64,26 @@
     @PostMapping(value = "/register", consumes = APPLICATION_JSON_VALUE)
     public void register(
             @RequestBody @Valid RegisterRequest registerRequest,
-            HttpServletRequest request
-    ) throws RecaptchaException {
+            HttpServletRequest request,
+            @RequestParam(required = false) String token
+    ) throws RecaptchaException, NotFoundException {
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java:69]: It won't throw NFE anymore. Remove it from the declaration.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java [hunk 6]
```diff
@@ -64,14 +64,26 @@
...
+
+            // IMPORTANT!!! not found invitation shouldn't return 404 for correct user registration
+            try {
+                if (token != null) {
+                    invitationService.confirmInvitation(email, token);
+                    val confirmInvitationEvent = new OnConfirmInvitationEvent(token, LocaleContextHolder.getLocale());
+                    eventPublisher.publishEvent(confirmInvitationEvent);
+                }
+            }
+            catch (NotFoundException e) {}
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java:85]: Catch Exception, and not NFE, as the called methods may throw runtime exceptions too.
I'd also log the thrown exceptions at level info.

Also, remove the newline between the closing try bracket and `catch`.

File: s4e-backend/src/test/java/pl/cyfronet/s4e/controller/AppUserControllerTest.java [hunk 7]
```diff
@@ -175,6 +173,68 @@ public void handle(OnResendRegistrationTokenEvent event) {
...
+                .email("some@email.pl")
+                .name("Name")
+                .surname("Surname")
+                .password("admin123")
+                .domain(AppUser.ScientificDomain.ATMOSPHERE)
+                .usage(AppUser.Usage.RESEARCH)
+                .country("PL")
+                .build();
+
+        assertThat(appUserRepository.findByEmail(registerRequest.getEmail()), isEmpty());
```
Comment (jswk) [inline s4e-backend/src/test/java/pl/cyfronet/s4e/controller/AppUserControllerTest.java:232]: Verify a user was created after the endpoint call.

**Review by jswk — CHANGES_REQUESTED** [review]

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java [hunk 8]
```diff
@@ -64,14 +64,27 @@
...
+
+            // IMPORTANT!!! not found invitation shouldn't return 404 for correct user registration
+            try {
+                if (token != null) {
+                    invitationService.confirmInvitation(email, token);
+                    val confirmInvitationEvent = new OnConfirmInvitationEvent(token, LocaleContextHolder.getLocale());
+                    eventPublisher.publishEvent(confirmInvitationEvent);
+                }
+            } catch (Exception e) {
+                val errorMessage = e.getMessage();
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java:85]: You must've not commited something when adding logging.

**Review by jswk — CHANGES_REQUESTED** [review]

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java [hunk 9]
```diff
@@ -64,14 +64,25 @@
...
             appUserService.register(registerRequest);
+
+            // IMPORTANT!!! not found invitation shouldn't return 404 for correct user registration
+            try {
+                if (token != null) {
+                    invitationService.confirmInvitation(email, token);
+                    val confirmInvitationEvent = new OnConfirmInvitationEvent(token, LocaleContextHolder.getLocale());
+                    eventPublisher.publishEvent(confirmInvitationEvent);
+                }
+            } catch (Exception e) {}
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java:84]: And why not log, as I suggested in https://github.com/cyfronet-fid/sat4envi/pull/888#discussion_r524011741?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java [hunk 10]
```diff
@@ -64,14 +64,27 @@
...
+
+            // IMPORTANT!!! not found invitation shouldn't return 404 for correct user registration
+            try {
+                if (token != null) {
+                    invitationService.confirmInvitation(email, token);
+                    val confirmInvitationEvent = new OnConfirmInvitationEvent(token, LocaleContextHolder.getLocale());
+                    eventPublisher.publishEvent(confirmInvitationEvent);
+                }
+            } catch (Exception e) {
+
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java:85]: ?

**Review by jswk — CHANGES_REQUESTED** [review]

**Review by kmarszalek — APPROVED** [review]

**Review by jswk — APPROVED** [review]

**Pull request merged**